l2geth/rollup/sync_service.go

```diff
@@ -763,10 +680,211 @@ func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("Cannot get latest transaction batch: %w", err)
+		}
+
+		nextIndex := s.GetNextBatchIndex()
+		if nextIndex-1 == latest.Index {
+			log.Info("No new batches to sync", "tip", latest.Index)
+			return nil
+		}
+		if err := s.syncTransactionBatchRange(nextIndex, latest.Index); err != nil {
+			return err
+		}
+
+		post, _, err := s.client.GetLatestTransactionBatch()
+		if err != nil {
+			return fmt.Errorf("Cannot get latest transaction batch: %w", err)
+		}
+		if post.Index == latest.Index {
+			return nil
+		}
+	}
+}
+
+// syncTransactionBatchRange will sync a range of batched transactions from
+// start to end (inclusive)
+func (s *SyncService) syncTransactionBatchRange(start, end uint64) error {
+	log.Debug("Syncing transaction batch range", "start", start, "end", end)
+	for i := start; i <= end; i++ {
+		log.Debug("Fetching transaction batch", "index", i)
+		_, txs, err := s.client.GetTransactionBatch(i)
+		if err != nil {
+			return fmt.Errorf("Cannot get transaction batch: %w", err)
+		}
+		for _, tx := range txs {
+			if err := s.applyBatchedTransaction(tx); err != nil {
+				return fmt.Errorf("cannot apply batched transaction: %w", err)
+			}
+		}
+		s.SetLatestBatchIndex(&i)
+	}
+	return nil
+}
+
+// syncQueueToTip will sync the `enqueue` transactions to the tip
+// from the last known `enqueue` transaction
+func (s *SyncService) syncQueueToTip() error {
+	s.loopLock.Lock()
+	defer s.loopLock.Unlock()
+
+	for {
+		latest, err := s.client.GetLatestEnqueue()
+		if errors.Is(err, errElementNotFound) {
+			log.Info("No queue transactions to sync")
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("Cannot get latest enqueue transaction: %w", err)
+		}
+		latestIndex := latest.GetMeta().QueueIndex
+		if latestIndex == nil {
+			return errors.New("Latest queue transaction has no queue index")
+		}
+
+		nextIndex := s.GetNextEnqueueIndex()
+		if nextIndex-1 == *latestIndex {
+			log.Info("No new queue transactions to sync", "tip", *latestIndex)
+			return nil
+		}
+		if err := s.syncQueueTransactionRange(nextIndex, *latestIndex); err != nil {
+			return err
+		}
+
+		post, err := s.client.GetLatestEnqueue()
+		if err != nil {
+			return fmt.Errorf("Cannot get latest transaction: %w", err)
+		}
+		postLatestIndex := post.GetMeta().QueueIndex
+		if postLatestIndex == nil {
+			return errors.New("Latest queue index is nil")
+		}
+		if *latestIndex == *postLatestIndex {
+			return nil
+		}
+	}
+}
+
+// syncQueueTransactionRange will apply a range of queue transactions from
+// start to end (inclusive)
+func (s *SyncService) syncQueueTransactionRange(start, end uint64) error {
+	log.Info("Syncing enqueue transactions to tip", "start", start, "end", end)
+	for i := start; i <= end; i++ {
+		tx, err := s.client.GetEnqueue(i)
+		if err != nil {
+			return fmt.Errorf("Canot get enqueue transaction; %w", err)
+		}
+		if err := s.applyTransaction(tx); err != nil {
```

Comment:
Should there be any validation on the transaction here, similar to the `ValidateAndApplySequencerTransaction`?

Reply:
The `enqueue` transactions are validated by the L1 contracts so we can assume that they have already been validated by this point